nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java

```diff
@@ -714,17 +750,90 @@ private void generateConstraintsForParam(
+   *
+   * @param argumentType the original type of the argument
+   * @param argument the argument expression tree
+   * @param state the visitor state
+   * @param path the tree path to the invocation if available and possibly distinct from {@code
+   *     state.getPath()}
+   * @return the refined type of the argument
+   */
+  private Type refineArgumentTypeWithDataflow(
+      Type argumentType, ExpressionTree argument, VisitorState state, @Nullable TreePath path) {
+    if (argumentType.isPrimitive()) {
+      return argumentType;
+    }
+    Symbol argumentSymbol = ASTHelpers.getSymbol(argument);
+    if (argumentSymbol == null) {
+      return argumentType;
+    }
+    TreePath currentPath = path != null ? path : state.getPath();
+    // this is a bit sketchy, as it may not actually be the valid tree path to the argument.
+    // However, all we need the path for is to discover the enclosing method/lambda/initializer, and
+    // for that purpose this should be sufficient.
+    TreePath argumentPath = new TreePath(currentPath, argument);
+    if (NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(argumentPath) == null) {
+      return argumentType;
+    }
+    Nullness refinedNullness;
+    if (calledFromDataflow) {
+      // dataflow analysis is already running, so just get the current dataflow value for the
+      // argument
+      refinedNullness =
+          analysis.getNullnessAnalysis(state).getNullnessFromRunning(argumentPath, state.context);
+    } else {
+      refinedNullness =
+          analysis.getNullnessAnalysis(state).getNullness(argumentPath, state.context);
+    }
+    if (refinedNullness == null) {
+      return argumentType;
+    }
+    return updateTypeWithNullness(state, argumentType, refinedNullness);
+  }
+
+  private Type updateTypeWithNullness(
+      VisitorState state, Type argumentType, Nullness refinedNullness) {
+    if (NullabilityUtil.nullnessToBool(refinedNullness)) {
+      // refine to @Nullable
+      if (isNullableAnnotated(argumentType)) {
+        return argumentType;
+      }
+      return TypeSubstitutionUtils.typeWithAnnot(argumentType, getSyntheticNullAnnotType(state));
+    } else {
+      // refine to @NonNull, by removing the top-level @Nullable annotation if present.
+      if (!isNullableAnnotated(argumentType)) {
+        return argumentType;
+      }
+      ListBuffer<Attribute.TypeCompound> updatedAnnotations = new ListBuffer<>();
+      boolean removedNullable = false;
+      for (Attribute.TypeCompound annot : argumentType.getAnnotationMirrors()) {
+        String annotationName = annot.type.toString();
+        if (Nullness.isNullableAnnotation(annotationName, config)) {
+          removedNullable = true;
+          continue;
+        }
+        updatedAnnotations.append(annot);
+      }
+      Verify.verify(removedNullable);
+      return TYPE_METADATA_BUILDER.cloneTypeWithMetadata(
+          argumentType, TYPE_METADATA_BUILDER.create(updatedAnnotations.toList()));
```

Comment:
Nit/optional: Should we make a `TypeSubstitutionUtils.typeMinusAnnot(...)` or even `TypeSubstitutionUtils.removeNullableAnnotions(...)` and move this logic there for symmetry? Not sure of the layering between `TypeSubstitutionUtils` and `Nullness`, but otherwise this doesn't seem to depend on anything in `GenericsChecks` but the type and `Config`.